hw/mcu/nxp/kinetis/src/hal_lpuart.c

```diff
@@ -0,0 +1,478 @@
+        return;
+    }
+    u = &uarts[port];
+    if (!u->u_configured || !u->u_open) {
+        return;
+    }
+
+    u->u_rx_stall = 0;
+
+    /* Send back what's in the RX ring buffer until it's empty or we get an
+     * error */
+    while ((rc >= 0) && !ur_is_empty(&u->ur_rx)) {
+        OS_ENTER_CRITICAL(sr);
+        rc = u->u_rx_func(u->u_func_arg, ur_read(&u->ur_rx));
+        if (rc >= 0) {
+            ur_bump(&u->ur_rx);
+        } else {
+            u->u_rx_stall = 1;
+        }
+        OS_EXIT_CRITICAL(sr);
+    }
+}
+
+static void
+uart_irq_handler(int port)
+{
+    struct hal_uart *u;
+    uint32_t status;
+    uint8_t data;
+
+    u = &uarts[port];
+    if (u->u_configured && u->u_open) {
+        status = LPUART_GetStatusFlags(u->u_base);
+        /* Check for RX data */
+        if (status & (kLPUART_RxDataRegFullFlag | kLPUART_RxOverrunFlag)) {
+            data = LPUART_ReadByte(u->u_base);
+            if (u->u_rx_stall || u->u_rx_func(u->u_func_arg, data) < 0) {
+                /*
+                 * RX queue full.
+                 */
+                u->u_rx_stall = 1;
+                ur_queue(&u->ur_rx, data);
+            }
+        }
+        /* Check for TX complete */
+        if (kLPUART_TxDataRegEmptyFlag & LPUART_GetStatusFlags(u->u_base)) {
+            if (u->u_tx_started) {
+                u->u_tx_started = 0;
+                if (u->u_tx_done) {
+                    u->u_tx_done(u->u_func_arg);
+                }
+            }
+        }
+    }
+}
+
+static void
+uart_irq0(void)
+{
+    uart_irq_handler(0);
+}
+
+static void
+uart_irq1(void)
+{
+    uart_irq_handler(1);
+}
+
+static void
+uart_irq2(void)
+{
+    uart_irq_handler(2);
+}
+
+static void
+uart_irq3(void)
+{
+    uart_irq_handler(3);
+}
+
+static void
+uart_irq4(void)
+{
+    uart_irq_handler(4);
+}
+
+int
+hal_uart_config(int port,
+                int32_t speed,
+                uint8_t databits,
+                uint8_t stopbits,
+                enum hal_uart_parity parity,
+                enum hal_uart_flow_ctl flow_ctl)
+{
+    struct hal_uart *u;
+    lpuart_config_t uconfig;
+
+    if (port >= FSL_FEATURE_SOC_LPUART_COUNT) {
+        return -1;
+    }
+    u = &uarts[port];
+    if (!u->u_configured || u->u_open) {
+        return -1;
+    }
+
+    /* PIN config (all UARTs use kPORT_MuxAlt3) */
+    CLOCK_EnableClock(u->p_clock);
+    PORT_SetPinMux(u->p_base, u->u_pin_rx, kPORT_MuxAlt3);
+    PORT_SetPinMux(u->p_base, u->u_pin_tx, kPORT_MuxAlt3);
+
+    /* UART CONFIG */
+    CLOCK_SetLpuartClock(2U);
+
+    LPUART_GetDefaultConfig(&uconfig);
+    uconfig.baudRate_Bps = speed;
+
+    switch (databits) {
+    case 8:
+        uconfig.dataBitsCount = kLPUART_EightDataBits;
+        break;
+    case 7:
+#if FSL_FEATURE_LPUART_HAS_7BIT_DATA_SUPPORT
+        uconfig.dataBitsCount = kLPUART_SevenDataBits;
+        break;
+#endif /* Fallthrought */
+    default:
+        uconfig.parityMode = kLPUART_ParityEven;
+        return -1;
```

Comment:
No reason to set `partityMode` if you're going to leave early right?

Reply:
hah good catch!

Reply:
Addressed.